File: gems/decomplex/lib/decomplex/temporal_ordering_pressure.rb
# frozen_string_literal: true

require_relative "syntax"

module Decomplex
  # TemporalOrderingPressure -- classes/modules whose public method
  # surface exposes a mutable lifecycle. If several public methods read
  # or write the same instance state, callers can invoke those methods
  # in many orders and create an implicit state machine.
  class TemporalOrderingPressure
    MethodState = Struct.new(:name, :line, :span, :visibility, :reads, :writes,
                             keyword_init: true)

    def self.scan(files)
      rows = files.flat_map do |file|
        document = Syntax.parse(file, parser: "tree_sitter")
        new(file, document).scan
      end
      rows.sort_by { |h| [-h[:score], -h[:state_methods], h[:file], h[:owner]] }
    end

    def initialize(file, document)
      @file = file
      @document = document
    end

    def scan
      temporal_owners.filter_map do |owner|
        row = pressure_row(owner, owner_methods(owner))
        row if row
      end
    end

    private

    def temporal_owners
      (@document.owner_defs.map(&:name) + @document.function_defs.map(&:owner)).compact.uniq
    end

    def owner_methods(owner)
      @document.function_defs.select { |function| function.owner == owner }.map do |function|
        MethodState.new(
          name: function.name,
          line: function.line,
          span: function.span,
          visibility: function.visibility || :public,
          reads: state_reads_for(function).uniq.sort,
          writes: state_writes_for(function).uniq.sort
        )
      end
    end

    def state_reads_for(function)
      @document.state_reads.select do |read|
        read.owner == function.owner && read.function == function.name
      end.map(&:field)
    end

    def state_writes_for(function)
      @document.state_writes.select do |write|
        write.owner == function.owner && write.function == function.name
      end.map(&:field)
    end

    def pressure_row(owner, methods)
      public_methods = methods.select { |m| m.visibility == :public }
      state_methods = public_methods.select { |m| !(m.reads + m.writes).empty? }
      writers = public_methods.select { |m| !m.writes.empty? }
      return nil if state_methods.size < 3 || writers.size < 2

      fields = state_methods.flat_map { |m| m.reads + m.writes }.uniq.sort
      shared_fields = fields.select do |field|
        state_methods.count { |m| (m.reads + m.writes).include?(field) } >= 2
      end
      return nil if shared_fields.empty?

      n = state_methods.size
      state_space = 2**[fields.size, 12].min
      score = (n * writers.size * [shared_fields.size, 1].max) + state_space
      {
        at: "#{@file}:#{owner}:#{state_methods.first.line}",
        file: @file,
        owner: owner,
        public_methods: public_methods.size,
        state_methods: n,
        writers: writers.size,
        state_fields: fields,
        shared_fields: shared_fields,
        orderings: factorial_label(n),
        state_space: "2^#{fields.size}",
        score: score,
        sites: state_methods.map { |m| "#{@file}:#{m.name}:#{m.line}" },
        spans: state_methods.to_h { |m| ["#{@file}:#{m.name}:#{m.line}", m.span] }
      }
    end

    def factorial_label(n)
      "#{n}!"
    end
  end
end

File: gems/decomplex/rust/src/decomplex/detectors/temporal_ordering_pressure.rs
use crate::decomplex::syntax::{self, Document, Language, Span};
use anyhow::Result;
use serde::Serialize;
use std::collections::{BTreeMap, BTreeSet};
use std::path::PathBuf;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct TemporalOrderingPressureRow {
    pub at: String,
    pub file: String,
    pub owner: String,
    pub public_methods: usize,
    pub state_methods: usize,
    pub writers: usize,
    pub state_fields: Vec<String>,
    pub shared_fields: Vec<String>,
    pub orderings: String,
    pub state_space: String,
    pub score: usize,
    pub sites: Vec<String>,
    pub spans: BTreeMap<String, Span>,
}

#[derive(Clone, Debug)]
struct MethodState {
    name: String,
    line: usize,
    span: Span,
    visibility: String,
    reads: Vec<String>,
    writes: Vec<String>,
}

pub fn scan_files(
    files: &[PathBuf],
    language: Language,
) -> Result<Vec<TemporalOrderingPressureRow>> {
    let documents = syntax::parse_files(files, language)?;
    Ok(scan_documents(&documents))
}

pub fn scan_documents(documents: &[Document]) -> Vec<TemporalOrderingPressureRow> {
    let mut rows = Vec::new();
    for document in documents {
        rows.extend(scan_document_facts(document));
    }
    rows.sort_by(|a, b| {
        b.score
            .cmp(&a.score)
            .then_with(|| b.state_methods.cmp(&a.state_methods))
            .then_with(|| a.file.cmp(&b.file))
            .then_with(|| a.owner.cmp(&b.owner))
    });
    rows
}

fn scan_document_facts(document: &Document) -> Vec<TemporalOrderingPressureRow> {
    let owners = document
        .owner_defs
        .iter()
        .map(|owner| owner.name.clone())
        .chain(
            document
                .function_defs
                .iter()
                .map(|function| function.owner.clone()),
        )
        .filter(|owner| !owner.is_empty())
        .collect::<BTreeSet<_>>();
    owners
        .into_iter()
        .filter_map(|owner| pressure_row_for_owner(document, &owner))
        .collect()
}

fn pressure_row_for_owner(document: &Document, owner: &str) -> Option<TemporalOrderingPressureRow> {
    let methods = document
        .function_defs
        .iter()
        .filter(|function| function.owner == owner)
        .map(|function| MethodState {
            name: function.name.clone(),
            line: function.line,
            span: function.span,
            visibility: function
                .visibility
                .clone()
                .unwrap_or_else(|| "public".to_string()),
            reads: sorted_unique(
                document
                    .state_reads
                    .iter()
                    .filter(|read| read.owner == function.owner && read.function == function.name)
                    .map(|read| read.field.clone()),
            ),
            writes: sorted_unique(
                document
                    .state_writes
                    .iter()
                    .filter(|write| {
                        write.owner == function.owner && write.function == function.name
                    })
                    .map(|write| write.field.clone()),
            ),
        })
        .collect::<Vec<_>>();
    pressure_row(document.file.as_str(), owner, &methods)
}

fn pressure_row(
    file: &str,
    owner: &str,
    methods: &[MethodState],
) -> Option<TemporalOrderingPressureRow> {
    let public_methods: Vec<_> = methods
        .iter()
        .filter(|m| m.visibility == "public")
        .collect();
    let state_methods: Vec<_> = public_methods
        .iter()
        .filter(|m| !m.reads.is_empty() || !m.writes.is_empty())
        .collect();
    let writers: Vec<_> = public_methods
        .iter()
        .filter(|m| !m.writes.is_empty())
        .collect();

    if state_methods.len() < 3 || writers.len() < 2 {
        return None;
    }

    let mut fields_set = BTreeSet::new();
    for m in &state_methods {
        fields_set.extend(m.reads.iter().cloned());
        fields_set.extend(m.writes.iter().cloned());
    }
    let fields = fields_set.into_iter().collect::<Vec<_>>();
    let shared_fields = fields
        .iter()
        .filter(|field| {
            state_methods
                .iter()
                .filter(|m| m.reads.contains(*field) || m.writes.contains(*field))
                .count()
                >= 2
        })
        .cloned()
        .collect::<Vec<_>>();
    if shared_fields.is_empty() {
        return None;
    }

    let n = state_methods.len();
    let state_space = 2usize.pow(fields.len().min(12) as u32);
    let score = (n * writers.len() * shared_fields.len().max(1)) + state_space;
    Some(TemporalOrderingPressureRow {
        at: format!("{}:{}:{}", file, owner, state_methods[0].line),
        file: file.to_string(),
        owner: owner.to_string(),
        public_methods: public_methods.len(),
        state_methods: n,
        writers: writers.len(),
        state_fields: fields,
        shared_fields,
        orderings: format!("{n}!"),
        state_space: format!(
            "2^{}",
            state_methods
                .iter()
                .flat_map(|m| m.reads.iter().chain(m.writes.iter()))
                .collect::<BTreeSet<_>>()
                .len()
        ),
        score,
        sites: state_methods
            .iter()
            .map(|m| format!("{}:{}:{}", file, m.name, m.line))
            .collect(),
        spans: state_methods
            .iter()
            .map(|m| (format!("{}:{}:{}", file, m.name, m.line), m.span))
            .collect(),
    })
}

fn sorted_unique(values: impl Iterator<Item = String>) -> Vec<String> {
    let mut out: Vec<_> = values.collect::<BTreeSet<_>>().into_iter().collect();
    out.sort();
    out
}
